ipwb/replay.py

```diff
@@ -179,6 +180,107 @@ def showMementosForURIRs_sansJS():
     return redirect('/memento/*/' + urir, code=301)
 
 
+def bin_search(iter, key, datetime=None):
+    # Read line encompassing current position
+    ln = iter.readline()
+
+    while ipwbUtils.isCDXJMetadataRecord(ln):
+        ln = iter.readline()
+
+    surtk, datetimeK, rest = ln.split(maxsplit=2)
+
+    matchDegree = getMatchDegree(surt, datetime, surtk, datetimeK)
+    if matchDegree == MementoMatch.EXACTMATCH:
+        return [ln]
+
+    # If further searching required...
+    left = 0
```

Comment:
This initial section has a few issues:

* I the `RIGHTKEYWRONGDATE` match occurs at the first data line then it will not be included in the response as the while loop below will already advance to the second data line.
* However, it may not be happening because the last line in the select sets the `left` to `0` (i.e., brings the start of the search to the beginning), which means all this metadata skipping effort is rendered useless.
* We really do not need to worry about skipping metadata lines as such because for a usual SURT key those lines will not match any way.
* We should also add an exception handling for `ValueError: not enough values to unpack (expected 3, got 2)` when performing split and assigning it to a fixed number of variables, because some metadata lines might not have two spaces in them, if we chose to perform bin search on the whole file without filtering any metadata lines out.

That said, following changes should allow skipping metadata lines and setting the start boundary properly:

```suggestion
    # Skip metadata lines
    while iter.peek(1)[:1] == b'!':
        iter.readline()

    # Set the beginning position to the start of the first data line
    left = iter.tell()
```

Reply:
Applied these in https://github.com/oduwsdl/ipwb/tree/issue-604-metadata for tests to be run.

Reply:
A couple of the tests in TravisCI failed with this change, reporting the unexpected error code. This will be a good opportunity to verify the expected status code and that the revised binsearch algorithm is returning the expected representation.

Reply:
I think we should focus on unit tests for this binsearch implementation to catch issues around corner cases, instead of integration test.

Reply:
I agree, but these integration tests should pass, too. Perhaps some unit tests would help remedy the issues. It might help with coverage, too.

Reply:
@ibnesayeed In planning these unit tests, should the data be within the test code or read from a file? I foresee enumerating all of the permutations for testing `bin_search` could make for some hard-to-maintain test code at the same time, it would best test the function itself to have the data inline.

Reply:
> I agree, but these integration tests should pass, too. Perhaps some unit tests would help remedy the issues. It might help with coverage, too.

Integration tests are useful and they should pass. They catch many issues, but they are generally not very effective in isolating the root causes.

Reply:
`bin_search` expects a file handle with its usage of `readline()`, so mocking this with converting a list of strings converted to an iterator might not work...unless there is a way to bind the iterator's `readline()` function (which does not pre-exist) to `next()`.

Reply:
> In planning these unit tests, should the data be within the test code or read from a file? I foresee enumerating all of the permutations for testing bin_search could make for some hard-to-maintain test code at the same time, it would best test the function itself to have the data inline.

I would create an index file that is only geared towards binary search test, not for replay. Then I will place data suitable for various scenarios in it and device test cases against this. I will place unique information in the JSON block to identify the line(s) that should show up (or not show up) in the result of various test cases. Following is a sample that I have put together very quickly, but more lines can be added to it without altering the essence of the existing lines. It is likely that a single file might not be sufficient to test every scenario (for example, if we were to test how will the bin search behave in an empty file or one with a single line), but it will be a good start.

```
!context ["http://oduwsdl.github.io/contexts/memento"]
!meta {"created_at": "2020-03-03T19:13:37", "purpose": "BinarySearch Testing"}
com,cnn)/ 20080117022613 {"desc": "first line"}
com,cnn)/ 20100325193658 {"desc": "second line"}
com,cnn)/ 20100327172049 {"desc": "third line"}
com,cnn)/ 20121114231445 {"desc": "long line", "lorem": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi ullamcorper diam dolor, sed molestie nisl iaculis vitae. Ut faucibus magna dapibus tortor scelerisque ullamcorper. Aenean varius fermentum lorem id placerat. Praesent sed luctus eros. Vestibulum facilisis tristique fermentum. Integer et tempus nunc. Nunc luctus dolor id risus dignissim, a maximus risus facilisis. Nam nec purus malesuada, viverra neque vel, luctus lacus. Phasellus facilisis tempus congue. Maecenas posuere tincidunt nisl, eu vestibulum nisl lobortis ut. Vestibulum porta elit sodales, laoreet dolor quis, volutpat quam. Cras tincidunt imperdiet lacus sit amet rutrum. Sed posuere, ipsum et pretium tempus, orci ex pellentesque ante, eu aliquet augue velit bibendum justo."}
com,cnn)/ 20181221190104 {"desc": "last homepage line"}
com,cnn)/news 20170111124504 {"desc": "single capture line"}
com,cnn)/news?id=ipwb-is-awesome&type=latest+news 20000306162047 {"desc": "line with query param"}
uk,co,bbc)/ 20051222010821 {"desc": "root path"}
uk,co,bbc)/ 20120320091103 {"desc": "first duplicate"}
uk,co,bbc)/ 20120320091103 {"desc": "second duplicate"}
uk,co,bbc)/ 20190121071213 {"desc": "closest past to 20190123071550"}
uk,co,bbc)/ 20190123071559 {"desc": "closest to 20190123071550"}
uk,co,bbc)/ 20200111053342 {"desc": "last capture"}
xyz,example,foo)/bar/baz?blah 20071114090832 {"desc": "last line"}
```

Reply:
> `bin_search` expects a file handle with its usage of `readline()`, so mocking this with converting a list of strings converted to an iterator might not work...unless there is a way to bind the iterator's `readline()` function (which does not pre-exist) to `next()`.

This function is about performing searches in a file, not in the memory, so I would not even think about putting data inline, but a separate index file is what should be used as suggested above. That said, if there is a situation where you do need to mock inline data as a file object, you can use [io.BytesIO](https://docs.python.org/3/library/io.html#io.BytesIO).